controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+	// Nevertheless, we are checking here.  If this is an overkill, we can remove it.
+	secondary = a.ensureVRGIsSecondary(curHomeCluster)
+	if !secondary {
+		a.log.Info("Still waiting for VRG to transition to secondary", "clusterName", curHomeCluster)
+
+		return !ensured, nil
 	}
 
-	// Create manifestwork for all PVs for this subscription
-	return a.createOrUpdatePVsManifestWork(subscription.Name, subscription.Namespace, homeCluster, pvList)
+	return ensured, nil
 }
 
-func (a *AVRInstance) createOrUpdatePVsManifestWork(
-	name string, namespace string, homeClusterName string, pvList []corev1.PersistentVolume) error {
-	a.log.Info("Creating manifest work for PVs", "subscription",
-		name, "cluster", homeClusterName, "PV count", len(pvList))
+// runPlacementTast is a series of steps to creating, updating, and cleaning up
+// the necessary objects for the failover, failback, or relocation
+func (a *AVRInstance) runPlacementTask(targetCluster, targetClusterNamespace, oldHomeCluster string) (bool, error) {
+	// Restore PV to preferredCluster
+	err := a.createPVManifestWorkForRestore(targetCluster)
+	if err != nil {
+		return false, err
+	}
 
-	mwName := BuildManifestWorkName(name, namespace, MWTypePV)
+	err = a.updateUserPlRuleAndCreateVRGMW(targetCluster, targetClusterNamespace)
+	if err != nil {
+		return false, err
+	}
 
-	manifestWork, err := a.generatePVManifestWork(mwName, homeClusterName, pvList)
+	// 9. Attempt to delete VRG MW from failed cluster once the VRG state has changed to Secondary
+	// This is attempt to clean up is not guaranteed to complete at this stage. Deleting the old VRG
+	// requires guaranteeing that the VRG has transitioned to secondary.
+	return a.cleanup(oldHomeCluster)
+}
+
+func (a *AVRInstance) updateUserPlRuleAndCreateVRGMW(homeCluster, homeClusterNamespace string) error {
+	err := a.updateUserPlacementRule(homeCluster, homeClusterNamespace)
 	if err != nil {
 		return err
 	}
 
-	return a.createOrUpdateManifestWork(manifestWork, homeClusterName)
+	// Creating VRG ManifestWork will be running in parallel with ACM deploying app resources
+	err = a.createVRGManifestWork(homeCluster)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func (a *AVRInstance) unpauseSubscription(subscription *subv1.Subscription) error {
-	// Subscription might have been changed since we last read it.  Get the latest...
-	subLookupKey := types.NamespacedName{Name: subscription.Name, Namespace: subscription.Namespace}
-	latestSub := &subv1.Subscription{}
+func (a *AVRInstance) createPVManifestWorkForRestore(newPrimary string) error {
+	pvMWName := BuildManifestWorkName(a.instance.Name, a.instance.Namespace, MWTypePV)
 
-	err := a.reconciler.Get(a.ctx, subLookupKey, latestSub)
-	if err != nil {
-		return fmt.Errorf("failed to retrieve subscription %+v", subLookupKey)
+	existAndApplied, err := a.isManifestExistAndApplied(pvMWName, newPrimary)
```

Comment:
(question) Is there a point at which an older PV manifest exists? IOW, a newer PV was created, but older restore manifest is present and hence newer PV is missed?